Vendor

buhl opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
In  #1587 we talked about a way to handle vendor libraries.

## Description
This PR vendors https://github.com/brettlangdon/git-vendor to vendor libs into bash-it.
This PR also includes the preexec lib and fixes to the source where needed. It also
includes a new plugin that can ring the terminal bell `echo -e \a` when a long running
process finishes.

## Motivation and Context
As discussed in #1587 I would be nice to have a better controlled/versioned way of
including libraries in bash-it. There have been talks of `git subtrees` and loader
scripts to aid the loading of plugins. This PR includes these tings.

## How Has This Been Tested?
I have been running with the new preexec lib for over a year in a personal fork of bash-it
the git-vendor lib is only used be dev who wish to add or update libs. It wont have any
effect on regular users of bash-it

## Types of changes
- [x] New feature (non-breaking change which adds functionality)
I feel I need to document how to add or update vendored libs. Where would be the best
place for this?


## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] If I have added a new file, I also added it to ``clean_files.txt`` and formatted it using ``lint_clean_files.sh``.
- [x] I have added tests to cover my changes, and all the new and existing tests pass.

NoahGorny commented:
I am impressed by your efforts @buhl !
I agree that this needs to be documented somewhere. I guess the `docs/development.rst` can be a good location to include such information.

Also, if you add new plugins, please add then to `clean_files.txt` and lint them. I am unsure what to do about the vendored code, and we need to discuss whether we should lint it as well.

buhl commented:
Thanks @NoahGorny!
I will be sure to tick that last boxes in the PR. I just wanted a go-ahead before putting to much work into it.
I don't think it's a good idea to lint the vendored code. It's libs we depend on, not source code we maintain.

NoahGorny commented:
> Thanks @NoahGorny!
> I will be sure to tick that last boxes in the PR. I just wanted a go-ahead before putting to much work into it.
> I don't think it's a good idea to lint the vendored code. It's libs we depend on, not source code we maintain.

I am also unsure whether we need to vendor `git-vendor` as well. Regular users do not need it, so we might as well just document how to install and use it for developers, without actually vendoring it inside Bash-it.

buhl commented:
> > Thanks @NoahGorny!
> > I will be sure to tick that last boxes in the PR. I just wanted a go-ahead before putting to much work into it.
> > I don't think it's a good idea to lint the vendored code. It's libs we depend on, not source code we maintain.
> 
> I am also unsure whether we need to vendor `git-vendor` as well. Regular users do not need it, so we might as well just document how to install and use it for developers, without actually vendoring it inside Bash-it.

Sure, I guess that's a reasonable requirements for a developer.

cornfeedhobo commented:
I really like where this discussion has gone. I notice that we're replacing the preexec library with a forked, and presumably better, version. Could we add [erichs/composure](https://github.com/erichs/composure) to this PR as well?

buhl commented:
@cornfeedhobo I don't see why we couldn't update composure as well once we landed on how to vendor libs.
But I think I would be more comfortable doing that in a separate PR (which I will gladly do once this is settled).

cornfeedhobo commented:
@buhl My thinking was that since you're already doing two big changes, might as well as a third. In the opposite direction of breaking up significant changes, maybe we should update preexec in place, and then do vendoring then?

NoahGorny commented:
I think we should remove the vendoring of `git vendor`, CR and merge this, and then continue with another vendoring PR of composure

buhl commented:
@cornfeedhobo I am comfortable with the preexec changes for two reasons.
1) I have been using this since April and have had no issues.
2) Very little in the bash-it source code actually uses preexec.
Composure on the other hand is AFAIK used in all plugins, aliases and bash_it.sh so I feel that is a bigger change.
However I don't think the change will be a problem, I would just like to isolate it to it's own PR.

cornfeedhobo commented:
@buhl those are good reasons. Thanks again for your contribution!

buhl commented:
@NoahGorny @cornfeedhobo  Sorry for the rebase. I had to make sure it could be done to give more clean PR's
Turns out it's a bit messy to de rebasing with subtrees. It should however, only be necessary when preparing a PR for a clean merge.
But I do think we/I need to play a bit more around with subtree before adding it to bash-it.
I have also updated the documentation regarding this issue. But it seems the tests for these are failing. I will look into this.
Other than that I removed git-vendor and added some files to clean_files.txt and checked that my changes also worked with the barbuk theme as it used preexec

buhl commented:
I will try to make a test of the plugin I added as well.

buhl commented:
I don't know why the lint is failing. Running the same script locally works.

cornfeedhobo commented:
@buhl Hmm, barbuk should already be cleaned if you rebased off master correctly. Maybe rebase again?

The only thing I can guess is that you've run into the same `shfmt` bug a few of us have. Essentially, our choice of formatting flags is causing a bug that adds whitespace on the first run in places you would not expect, like in bracket tests (`[[]]`). I've filed a bug upstream, but for now you'll just have to run `shfmt` like a dozen times on all the files you are adding.

buhl commented:
Thanks @cornfeedhobo, It was an error on my part :facepalm:

buhl commented:
I think maybe I should rebase this branch into fewer commits to make it more neat.
But @NoahGorny and/or @cornfeedhobo  could I get some comments on the code and documentations?
I must admit that the trouble with rebasing a feature branch that is behind master makes me hesitate a little in recommending subtrees as the way to go for vendoring code.
I have also looked at https://github.com/ingydotnet/git-subrepo but AFAIK it doesn't support releases (as in tags) and it is overall a little more clumsy to use compared to git-vendor. However it creates a more clean git history :shrug:

cornfeedhobo commented:
@buhl no problem. I'll take a look this evening

nwinkler commented on docs/development.rst:
```diff
@@ -78,6 +79,35 @@ Having the order based on a numeric priority in a common directory allows for mo
+
+Vendored libs are external libraries, meaning source code not maintained by bash-it
+developers.
+They are ``git sutrees`` curated in the ``vendor/`` folder. To ease the work with git
```
* `subtrees` is missing a `b`.
* `bash-it` should be `Bash-it`.

buhl replied:
Great, I will get those fixed.

buhl replied:
done

nwinkler commented on plugins/available/cmd-returned-notify.plugin.bash:
```diff
@@ -0,0 +1,16 @@
+# shellcheck shell=bash
```
Wouldn't this rather belong in a separate PR?

buhl replied:
Maybe, this plugin was why i needed to update preexec in the first place so I thought I would sneak it in. :wink: 
I suppose it could be PR'ed afterwards as well. How ever I also needed to test that the updated preexec library could run more functions per event. 
I am good either way

NoahGorny replied:
I am okay with including it in this PR

nwinkler reviewed (commented):
Left a couple of minor comments...

NoahGorny commented on test/plugins/xterm.plugin.bats:
```diff
@@ -0,0 +1,54 @@
+  setup_test_fixture
+
+  # Copy the test fixture to the Bash-it folder
+  if command -v rsync &> /dev/null
```
use `command_exists`, so you wont need to `/dev/null` it

buhl replied:
done

NoahGorny commented on docs/development.rst:
```diff
@@ -78,6 +79,35 @@ Having the order based on a numeric priority in a common directory allows for mo
+Vendored libs are external libraries, meaning source code not maintained by Bash-it
+developers.
+They are ``git subtrees`` curated in the ``vendor/`` folder. To ease the work with git
+vendored libs as subtrees we use the `git-vendor <https://github.com/Tyrben/git-vendor>`_ tool.
```
you should add a reference on how to use `git-vendor`, probably can refer to https://github.com/brettlangdon/git-vendor#usage

NoahGorny replied:
you are referring here to `Tyrben` which is just a lone fork, better reference `brettlangdon`

NoahGorny reviewed (commented):
Small nits from me, otherwise LGTM

NoahGorny commented:
ping @buhl :smile:

buhl commented:
Sorry! My analog life called. I will try to fix the last few things and rebase the branch to make it merge ready tomorrow 😊

buhl commented:
I did a rebase/force push to make the history a little nicer. So if you have pulled my branch be sure to delete it locally before pulling again.
I added a little more documentation. Other than that I think it's good to go?

NoahGorny reviewed (commented):
@buhl I left you a small comment, other than it LGTM

BTW, I see the `git-vendor` is not actively maintained, this is quite concerning, but if the tool works, we can use it and just fork it if needed

buhl commented:
@NoahGorny Sorry, must have misser your comment. Will look at it tomorrow.

Regarding git vendor I referenced https://github.com/Tyrben/git-vendor because I looks like an actively maintained fork of the other repo. If the fork goes stale as well we can consider maintaining a fork our selfs.
But for now Tyrben is active in brettlangdon repos an seems to port pull requests to their own fork.

NoahGorny commented:
> @NoahGorny Sorry, must have misser your comment. Will look at it tomorrow.
> 
> Regarding git vendor I referenced https://github.com/Tyrben/git-vendor because I looks like an actively maintained fork of the other repo. If the fork goes stale as well we can consider maintaining a fork our selfs.
> But for now Tyrben is active in brettlangdon repos an seems to port pull requests to their own fork.

Oh, I see, so ignore my comment, and instead add a note about the fact that we use the maintained fork and not the original repo (going to the supplied link might be confusing otherwise, I got quite confused)

Anyway, well done! I hope we can merge it soon and include all vendored code in this format

buhl commented:
Hmm, themes/atomic/atomic.theme.bash and themes/base.theme.bash are failing the lint check. I did not touch these files but should I fix the errors in this PR?

buhl commented:
@NoahGorny I added some text about the different git vendor repos. I think I have resolved all pending questions so now there is only the linting errors left mentioned in my previous comment.

NoahGorny commented:
> Hmm, themes/atomic/atomic.theme.bash and themes/base.theme.bash are failing the lint check. I did not touch these files but should I fix the errors in this PR?

Running locally, I saw `shfmt` complaining about the `xterm` plugin only, I added a fix commit, and I'll review this again now

NoahGorny reviewed (approved):
@buhl I made a couple of change, but now its LGTM

I'll let @nwinkler @davidpfarrell @cornfeedhobo voice their opinions as this is a big change. If we merge this, we should move all of our vendored code to be used in this pattern. I will help us control it!

buhl commented:
Ah, I missed the linting errors in the xterm test because I saw errors in the two themes mentioned in an earlier comment.
Thanks for fixing them :smile: 

I will gladly help moving the other libraries to this vendor method if its accepted.

nwinkler commented:
Looks good to me!

NoahGorny commented:
Its about time we merge this, :partying_face: 
@buhl you are welcome to move more things into the new `vendor` directory :smile: 

I added a lint check of the `init.d` directory, as it is our code

buhl commented:
Sounds great :smile: 
I think I will have some time the upcoming days to vendor some libraries.
I have started a new issue #1818 to keep track of this process.
For now there is only only external library on the list
